### DISCUSSION TITLE
Problem with creating a generic when using a method that uses `Grouped`

### DISCUSSION THREAD

**Discussion by vincent-herlemont** [diesel-rs/diesel#3968] [Q&A]
I'm trying to make a generic to select all the posts (example taken from the official diesel site) that belong to a list of IDs. Here is what it looks like with concrete types (all works ✅ ):

```rust
    pub fn find_posts(&mut self, ids: Vec<i32>) -> Vec<models::Post> {
        schema::posts::table
            .filter(schema::posts::id.eq_any(ids))
            .load::<models::Post>(&mut self.conn)
            .unwrap()
    }
```

Now, I want to use generic types, first the IDs which I will name PKS (for primary keys).
Here is my attempt (not working ❌ ):

```rust
    pub fn find_posts<'query, PKS, PK>(&mut self, ids: PKS) -> Vec<models::Post>
    where
        PK: diesel::expression::AsExpression<diesel::sql_types::Integer> + IntoIterator<Item = PK>,
        PKS: IntoIterator<Item = PK>,
        schema::posts::table: diesel::query_dsl::methods::FilterDsl<PK>,
        diesel::dsl::Filter<schema::posts::table, ??  >: RunQueryDsl<SqliteConnection>,
    {
        schema::posts::table
            .filter(schema::posts::id.eq_any(ids.into_iter()))
            .load::<models::Post>(&mut self.conn)
            .unwrap()
    }
```

You can see my `??` because that's where I don't know what to put.

Related error:
```bash
error[E0599]: the method `load` exists for associated type `<table as FilterDsl<Grouped<In<id, Many<Integer, PK>>>>>::Output`, but its trait bounds were not satisfied
   --> src/lib.rs:102:14
    |
100 | /         schema::posts::table
101 | |             .filter(schema::posts::id.eq_any(ids.into_iter()))
102 | |             .load::<models::Post>(&mut self.conn)
    | |             -^^^^ method cannot be called due to unsatisfied trait bounds
```
This error happens because [`EqAny`](https://docs.rs/diesel/latest/diesel/dsl/type.EqAny.html) use `Grouped` and I cannot access the struct [`diesel::expression::Grouped`](https://github.com/diesel-rs/diesel/blob/a3858e9143eae05c04aaac61708b4c9ba147944a/diesel/src/expression/mod.rs#L33
) because it is private to the crate.

Do you have any ideas to put me on the right way?

**Comment by weiznich** [comment]
I generally advice not to write generic code involving diesel outside of fundamental library implementations. You end up with complicated hard to maintain generic code. It's usually easier to follow the patterns described in the "Composing Applications" guide. 

These types are private as we don't want to expose them as part of the stable API as their representation can change at any point. You usually want to use the type defs from `diesel::dsl` in their place. (Consult the documentation of the relevant methods for which type def is returned by which method).
